base_login_date_improvement/res_users.py

```
@@ -0,0 +1,137 @@
+    def _get_login_date(self, cr, uid, ids, name, args, context=None):
+        res = {}
+        user_login_obj = self.pool['res.users.login']
+        for user_id in ids:
+            login_ids = user_login_obj.search(
+                cr, uid, [('user_id', '=', user_id)], limit=1,
+                context=context)
+            if len(login_ids) == 0:
+                res[user_id] = False
+            else:
+                login = user_login_obj.browse(cr, uid, login_ids[0],
+                                              context=context)
+                res[user_id] = login.login_dt
+        return res
+
+    _columns = {
+        'login_date': fields.function(_get_login_date,
+                                      string='Latest connection',
+                                      type='date', select=1,
+                                      readonly=True, store=False,
+                                      nodrop=True),
+    }
+
+    # Re-defining the login function in order to use the new table
+    def login(self, db, login, password):
+        if not password:
+            return False
+        user_id = False
+        cr = pooler.get_db(db).cursor()
+        try:
+            # check if user exists
+            res = self.search(cr, SUPERUSER_ID, [('login', '=', login)])
+            if res:
+                user_id = res[0]
+                try:
+                    # check credentials
+                    self.check_credentials(cr, user_id, password)
+                except openerp.exceptions.AccessDenied:
+                    _logger.info("Login failed for db:%s login:%s", db, login)
+                    user_id = False
+
+                if user_id:
+                    try:
+                        update_clause = ('NO KEY UPDATE'
+                                         if cr._cnx.server_version >= 90300
+                                         else 'UPDATE')
+                        cr.execute("SELECT login_dt "
```

I'm not sure FOR UPDATE would be more robust, since 2 transactions doing a SELECT FOR NO KEY UPDATE will result in the same behavior (unable to lock for the 2nd one). On the other hand, no table uses res_users_login as a foreign key, so to simplify, I will remove the condition and use FOR UPDATE. 
Also, I do not think a more thorught UPSERT is needed too; the behavior in Odoo is to "fail silently" if the lock was not available, and any concurrency condition would only add one second to the login date.

I'll rework the mecanism to use the SELECT FOR UPDATE first, then UPDATE or INSERT if something was found or not, and use either the lock or the unicity constraint on the user_id to fail and rollback if the row was created/updated concurrently.